Add ColQwen2 to 🤗 transformers

tonywu71 opened the pull request:
# What does this PR do?

Add ColQwen2 in 🤗 `transformers`. ColQwen2 is a model that uses the [ColPali](https://doi.org/10.48550/arXiv.2407.01449) architecture with a Qwen2-VL backbone.

## Who can review?

- @yonigozlan
- @ArthurZucker 
- @Cyrilvallez

## Additional details

- This PR uses the new [Modular 🤗 transformers](https://huggingface.co/docs/transformers/main/en/modular_transformers#modular-transformers)
- The ColPali is mainly inspired from the [colpali-engine](https://github.com/illuin-tech/colpali) repository I'm maintaining with my co-authors. The initial code was taken from `colpali-engine==v0.3.6`.
- The newly converted model weights are stored in [`vidore/colqwen2-v1.0-hf`](https://huggingface.co/vidore/colqwen-v1.0-hf).

## Progress checklist

- [x]  (Optional) Understood the model’s theoretical aspects
- [x]  Prepared 🤗 Transformers dev environment
- [x]  Set up debugging environment of the original repository
- [x]  Created script that successfully runs the forward() pass using the original repository and checkpoint
- [x]  Successfully added the model skeleton to 🤗 Transformers
- [x]  Successfully converted original checkpoint to 🤗 Transformers checkpoint
- [x]  Successfully ran forward() pass in 🤗 Transformers that gives identical output to original checkpoint
- [x]  Finished model tests in 🤗 Transformers
- [x]  Successfully added tokenizer in 🤗 Transformers
- [x]  Run end-to-end integration tests
- [x]  Finished docs
- [x]  Uploaded model weights to the Hub
- [x]  Submitted the pull request
- [x]  (Optional) Added a demo notebook → can be found in https://github.com/tonywu71/colpali-cookbooks/pull/17

tonywu71 committed:
feat: add colqwen2 (wip)

ArthurZucker commented:
Feel free to ping us once this is ready for review!

ArthurZucker commented:
Feel free to ping @Cyrilvallez once this is ready for review! 🤗

tonywu71 committed:
tests: fix test_attention_outputs

tonywu71 committed:
tests: reduce hidden size to accelerate tests

tonywu71 committed:
tests: fix `test_attention_outputs` 🥳

tonywu71 committed:
fix: fix wrong parent class for `ColQwen2ForRetrievalOutput`

tonywu71 committed:
fix: minor typing and style changes

tonywu71 committed:
chore: run `make style`

tonywu71 committed:
feat: remove redundant `max_num_visual_tokens` attribute in `ColQwen2Processor`

tonywu71 committed:
tests: tweak comments

tonywu71 committed:
style: apply ruff formatter

tonywu71 committed:
feat: move default values for `visual_prompt_prefix` and `query_prefix`

tonywu71 committed:
docs: update ColQwen2 model card

tonywu71 committed:
docs: tweak model cards

tonywu71 committed:
docs: add required example config checkpoint

tonywu71 committed:
tests: update expected scores in integration test

tonywu71 committed:
docs: tweak quickstart snippets

yonigozlan commented on tests/models/colqwen2/test_processing_colqwen2.py:
```diff
@@ -0,0 +1,250 @@
+class ColQwen2ProcessorTest(ProcessorTesterMixin, unittest.TestCase):
+    processor_class = ColQwen2Processor
+
+    def setUp(self):
```
We are using `def setUpClass` and `tearDownClass` now to launch the setup only once for all tests. You can look in test_processing_colpali for the exact formatting

yonigozlan commented on tests/models/colqwen2/test_processing_colqwen2.py:
```diff
@@ -0,0 +1,250 @@
+        self.assertIsInstance(batch_feature["input_ids"], torch.Tensor)
+        self.assertEqual(batch_feature["input_ids"].shape[0], len(queries))
+
+        # The following tests are overwritten as ColQwen2Processor can only take one of images or text as input at a time
```
```suggestion
    # The following tests are overwritten as ColQwen2Processor can only take one of images or text as input at a time
```

yonigozlan commented on tests/models/colqwen2/test_processing_colqwen2.py:
```diff
@@ -0,0 +1,250 @@
+SAMPLE_VOCAB = get_tests_dir("fixtures/test_sentencepiece.model")
+
+
+@require_vision
```
let's have 
```suggestion
@require_torch
@require_vision
```
here and remove all other @require on the methods

yonigozlan reviewed (commented):
Hey @tonywu71 ! Thanks for contributing 🤗. Looks almost ready to go to me, I just pointed out a few nits to change

tonywu71 committed:
fix: address PR comments

tonywu71 committed:
tests: fix colqwen2 tests + tweak comment in colpali test

tonywu71 committed:
tests: unskip useful tests

tonywu71 committed:
fix: fix bug when `visual_prompt_prefix` or `query_prefix` is an empty string

tonywu71 commented on src/transformers/models/colqwen2/modular_colqwen2.py:
```diff
@@ -0,0 +1,304 @@
+        self.image_token = "<|image_pad|>" if not hasattr(tokenizer, "image_token") else tokenizer.image_token
+        self.video_token = "<|video_pad|>" if not hasattr(tokenizer, "video_token") else tokenizer.video_token
+
+        self.visual_prompt_prefix = (
+            visual_prompt_prefix
+            or "<|im_start|>user\n<|vision_start|><|image_pad|><|vision_end|>Describe the image.<|im_end|><|endoftext|>"
+        )
+        self.query_prefix = query_prefix or "Query: "
```
I wanted to do so but `make fix-copies` and `python utils/modular_model_converter.py --files_to_parse src/transformers/models/colqwen2/modular_colqwen2.py` are not happy with the long default `visual_prompt_prefix` (or maybe because it has a `\n` in it).

After converting the modular, I get:

```python
            visual_prompt_prefix (`str`, *optional*, defaults to `"<|im_start|>user
    <|vision_start|><|image_pad|><|vision_end|>Describe the image.<|im_end|><|endoftext|>"`): A string that gets tokenized and prepended to the image tokens.
            query_prefix (`str`, *optional*, defaults to `"Query: "`): A prefix to be used for the query.
```

After running `make fix-copies`, I get:

```diff
            visual_prompt_prefix (`str`, *optional*, defaults to `"<|im_start|>user
+ <|vision_start|><|image_pad|><|vision_end|>Describe the image.<|im_end|><|endoftext|>"`): <fill_docstring>
            query_prefix (`str`, *optional*, defaults to `"Query: "`): <fill_docstring>
    <|vision_start|><|image_pad|><|vision_end|>Describe the image.<|im_end|><|endoftext|>"`): A string that gets tokenized and prepended to the image tokens.
            query_prefix (`str`, *optional*, defaults to `"Query: "`): A prefix to be used for the query.
```

A quick workaround is to set the default to `None`, and to overwrite the `None` values with the defaults when necessary.

yonigozlan replied:
Ah I see, let's keep it like you did then

tonywu71 committed:
fix: fix ColPali outputs when `return_dict == False`

tonywu71 committed:
fix: fix issue with PaliGemma output not being a dict

tonywu71 committed:
docs: set default dtype to bfloat16 in quickstart snippets

tonywu71 committed:
fix: fix error when `return_dict=False` in ColPali and ColQwen2

tonywu71 committed:
tests: fix special tokens not being replaced in input_ids

tonywu71 committed:
style: fix lint

tonywu71 committed:
fix: `ColQwen2Processor`'s `padding_side` is now set from `processor_config.json`

tonywu71 committed:
fix: remove unused `padding_side` in ColQwen2 model

tonywu71 commented on src/transformers/models/colqwen2/modeling_colqwen2.py:
```diff
@@ -0,0 +1,358 @@
+            self.embedding_dim,
+        )
+        self._tied_weights_keys = ["vlm.model.embed_tokens.weight", "vlm.lm_head.weight"]
+        self.padding_side = "left"
```
It was a typo, thanks for the catch!

tonywu71 committed:
docs: update ColQwen2's model doc

yonigozlan reviewed (commented):
Nice thanks for iterating! I see two small things left to change then LGTM for me!

yonigozlan commented on src/transformers/models/colqwen2/modeling_colqwen2.py:
````diff
@@ -0,0 +1,362 @@
+        image_grid_thw: Optional[torch.LongTensor] = None,
+        cache_position: Optional[torch.LongTensor] = None,
+    ) -> Union[Tuple, ColQwen2ForRetrievalOutput]:  # noqa: F821
+        r"""
+        Returns:
+        ```"""
````
```suggestion
        r"""
        Returns:
        """
```

tonywu71 committed:
fix: fix harcoded vlm backbone class in ColQwen2Config

tonywu71 committed:
fix: remove `padding_side` from ColQwen2Processor as should fed from kwargs

tonywu71 committed:
docs: fix typo in model docstring

tonywu71 committed:
docs: add illuin mention in model docs

tonywu71 commented:
Hi @Cyrilvallez, can you review my PR please? 🤗  @yonigozlan has already thoroughly reviewed it, but very happy to get your feedbacks as well!

tonywu71 committed:
fix: let `padding_size` be handled by `tokenizer_config.json`

tonywu71 commented on src/transformers/models/colqwen2/modular_colqwen2.py:
```diff
@@ -0,0 +1,304 @@
+        )
+        self.query_prefix = query_prefix or "Query: "
+
+        self.tokenizer.padding_side = "left"
```
Update: after discussion with @yonigozlan, I have realized it makes much more sense to let `tokenizer_config.json` handle `padding_side`. I've just applied the necessary changes!

tonywu71 committed:
docs: add colpali reference url in colqwen2's model doc

tonywu71 committed:
docs: add Hf mention in model docs

tonywu71 committed:
docs: add late interaction mention in model docs

tonywu71 committed:
docs: tweak colqwen2 model doc

tonywu71 committed:
docs: update reference checkpoint for ColPali to v1.3

tonywu71 committed:
docs: simplify quickstart snippets

Cyrilvallez commented on src/transformers/models/colqwen2/modeling_colqwen2.py:
```diff
@@ -0,0 +1,362 @@
+    _no_split_modules = []
+
+    def _init_weights(self, module):
+        std = (
+            self.config.initializer_range
+            if hasattr(self.config, "initializer_range")
+            else self.config.vlm_config.initializer_range
+        )
```
Would return `AttributeError` here with a fresh Qwen2VLConfig, see previous comment.
Also, in general let's do `getattr(object, name, fallback)` instead of `if hasattr() - else`

Cyrilvallez commented on src/transformers/models/colqwen2/modeling_colqwen2.py:
```diff
@@ -0,0 +1,362 @@
+
+_CONFIG_FOR_DOC = "ColQwen2Config"
+
+COLQWEN2_START_DOCSTRING = r"""
+    This model inherits from [`PreTrainedModel`]. Check the superclass documentation for the generic methods the
+    library implements for all its model (such as downloading or saving, resizing the input embeddings, pruning heads
+    etc.)
+
+    This model is also a PyTorch [torch.nn.Module](https://pytorch.org/docs/stable/nn.html#torch.nn.Module) subclass.
+    Use it as a regular PyTorch Module and refer to the PyTorch documentation for all matter related to general usage
+    and behavior.
+
+    Parameters:
+        config ([`ColQwen2Config`]):
+            Model configuration class with all the parameters of the model. Initializing with a config file does not
+            load the weights associated with the model, only the configuration. Check out the
+            [`~PreTrainedModel.from_pretrained`] method to load the model weights.
+"""
```
Should be inherited from colpali

Cyrilvallez commented on src/transformers/models/colqwen2/modeling_colqwen2.py:
```diff
@@ -0,0 +1,362 @@
+@add_start_docstrings(
+    "The bare ColQwen2 model outputting raw hidden-states without any specific head on top.",
+    COLQWEN2_START_DOCSTRING,
+)
+class ColQwen2PreTrainedModel(PreTrainedModel):
+    config_class = ColQwen2Config
```
This is identical to ColPali, so let's simply inherit with modular!

Cyrilvallez commented on src/transformers/models/colqwen2/modeling_colqwen2.py:
```diff
@@ -0,0 +1,362 @@
+    attentions: Optional[Tuple[torch.FloatTensor]] = None
+
+
+COLQWEN2_FOR_RETRIEVAL_INPUT_DOCSTRING = r"""
+    Args:
+        input_ids (`torch.LongTensor` of shape `(batch_size, sequence_length)`):
+            Indices of input sequence tokens in the vocabulary. Padding will be ignored by default should you provide
+            it.
+            Indices can be obtained using [`AutoTokenizer`]. See [`PreTrainedTokenizer.encode`] and
+            [`PreTrainedTokenizer.__call__`] for details.
+            [What are input IDs?](../glossary#input-ids)
+        pixel_values (`torch.FloatTensor` of shape `(batch_size, num_channels, image_size, image_size)):
+            The tensors corresponding to the input images. Pixel values can be obtained using
+            [`AutoImageProcessor`]. See [`Qwen2VLImageProcessor.__call__`] for details. If none,
+            ColQwen2 will only process text (query embeddings).
+        attention_mask (`torch.Tensor` of shape `(batch_size, sequence_length)`, *optional*):
+            Mask to avoid performing attention on padding token indices. Mask values selected in `[0, 1]`:
+            - 1 for tokens that are **not masked**,
+            - 0 for tokens that are **masked**.
+            [What are attention masks?](../glossary#attention-mask)
+            Indices can be obtained using [`AutoTokenizer`]. See [`PreTrainedTokenizer.encode`] and
+            [`PreTrainedTokenizer.__call__`] for details.
+            If `past_key_values` is used, optionally only the last `decoder_input_ids` have to be input (see
+            `past_key_values`).
+            If you want to change padding behavior, you should read [`modeling_opt._prepare_decoder_attention_mask`]
+            and modify to your needs. See diagram 1 in [the paper](https://arxiv.org/abs/1910.13461) for more
+            information on the default strategy.
+            - 1 indicates the head is **not masked**,
+            - 0 indicates the head is **masked**.
+        output_attentions (`bool`, *optional*):
+            Whether or not to return the attentions tensors of all attention layers. See `attentions` under returned
+            tensors for more detail.
+        output_hidden_states (`bool`, *optional*):
+            Whether or not to return the hidden states of all layers. See `hidden_states` under returned tensors for
+            more detail.
+        return_dict (`bool`, *optional*):
+            Whether or not to return a [`~utils.ModelOutput`] instead of a plain tuple.
+        kwargs (`Dict[str, Any]`, *optional*):
+            Additional key word arguments passed along to the vlm backbone model.
+"""
```
Similarly this is not needed if using proper inheritance from ColPali 🤗

tonywu71 replied:
Agreed (if we decide going for modular is a good idea)!

Cyrilvallez commented on src/transformers/models/colqwen2/modeling_colqwen2.py:
```diff
@@ -0,0 +1,362 @@
+    [*ColPali: Efficient Document Retrieval with Vision Language Models*](https://arxiv.org/abs/2407.01449).
+    """
+)
+class ColQwen2ForRetrieval(ColQwen2PreTrainedModel):
+    main_input_name = "input_ids"
+
+    def __init__(self, config: ColQwen2Config):
+        super().__init__(config)
+        self.config = config
+        self.vocab_size = config.vlm_config.vocab_size
+        self.vlm = AutoModelForImageTextToText.from_config(config.vlm_config)
+        self.embedding_dim = self.config.embedding_dim
+        self.embedding_proj_layer = nn.Linear(
+            self.config.vlm_config.hidden_size,
+            self.embedding_dim,
+        )
+        self._tied_weights_keys = ["vlm.model.embed_tokens.weight", "vlm.lm_head.weight"]
+        self.post_init()
+
+    def get_input_embeddings(self):
+        return self.vlm.model.embed_tokens
+
+    def set_input_embeddings(self, value):
+        self.vlm.model.embed_tokens = value
+
+    def get_output_embeddings(self):
+        return self.vlm.lm_head
+
+    def set_output_embeddings(self, new_embeddings):
+        self.vlm.lm_head = new_embeddings
+
+    def set_decoder(self, decoder):
+        self.vlm.model = decoder
+
+    def get_decoder(self):
+        return self.vlm.model
```
Let's inherit from ColPali in modular to avoid all of that

tonywu71 replied:
A bit similarly to what I've said in https://github.com/huggingface/transformers/pull/35778#discussion_r2055983686, Qwen2VL has an atypical VLM implementation. In particular, it doesn't have a `language_model` (contrarily to PaliGemma, see below). Thus, I have to override these methods in your comments anyway!

https://github.com/huggingface/transformers/blob/8ab296501a2e4c6ddada77118acd29c7c61c7795/src/transformers/models/paligemma/modeling_paligemma.py#L283-L290

Cyrilvallez commented on src/transformers/models/colqwen2/modeling_colqwen2.py:
```diff
@@ -0,0 +1,362 @@
+        output_hidden_states: Optional[bool] = None,
+        return_dict: Optional[bool] = None,
+        pixel_values: Optional[torch.Tensor] = None,
+        image_grid_thw: Optional[torch.LongTensor] = None,
+        cache_position: Optional[torch.LongTensor] = None,
+    ) -> Union[Tuple, ModelOutput]:
+        """
+        Forward through the Qwen2VL backbone.
+
+        Uses a custom forward method to fix an issue with the gradient flow when training with multiple GPUs. Code is
+        mostly copied from [`Qwen2VLForConditionalGeneration.forward`], excluding the code for video processing.
+        """
+        if inputs_embeds is None:
+            inputs_embeds = self.vlm.model.embed_tokens(input_ids)
+
+            if pixel_values is not None:
+                pixel_values = pixel_values.type(self.vlm.visual.get_dtype())
+                image_embeds = self.vlm.visual(pixel_values, grid_thw=image_grid_thw)
+                image_mask = (
+                    (input_ids == self.config.vlm_config.image_token_id).unsqueeze(-1).expand_as(inputs_embeds)
+                )
+                image_embeds = image_embeds.to(inputs_embeds.device, inputs_embeds.dtype)
+                inputs_embeds = inputs_embeds.masked_scatter(image_mask, image_embeds)
+
+            if attention_mask is not None:
+                attention_mask = attention_mask.to(inputs_embeds.device)
+
+        outputs = self.vlm.model(
+            input_ids=None,
+            position_ids=position_ids,
+            attention_mask=attention_mask,
+            past_key_values=past_key_values,
+            inputs_embeds=inputs_embeds,
+            use_cache=use_cache,
+            output_attentions=output_attentions,
+            output_hidden_states=output_hidden_states,
+            return_dict=return_dict,
+            cache_position=cache_position,
+        )
+        return outputs
```
Here this is quite brittle to use the direct name attributes as they may change for different backbones - better to use the model's forward, and can probably be removed and everything inherited from ColPali!

tonywu71 replied:
Agreed, but the thing is that I have specifically chosen not to make ColQwen2 inherit from ColPali because of this custom `inner_forward` method.

To be more precise, this `inner_forward` method is here because Qwen2VL uses some kind of sequence padding strategy for more efficient inference. However, this strategy is not suited for the contrastive learning used in ColQwen2 as we have encountered issues with gradient flowing with our multi-GPU DDP training.

To conclude, I think it's best not to use modular for `ColPali→ColQwen2` because of this special inference tweak. Wdyt?

Cyrilvallez commented on src/transformers/models/colqwen2/modular_colqwen2.py:
```diff
@@ -0,0 +1,304 @@
+# coding=utf-8
+# Copyright 2024 The HuggingFace Inc. team.
```
2025!

Cyrilvallez commented on src/transformers/models/colqwen2/modular_colqwen2.py:
```diff
@@ -0,0 +1,304 @@
+
+        if text is None and images is None:
+            raise ValueError("Either text or images must be provided")
+        if text is not None and images is not None:
+            raise ValueError("Only one of text or images can be processed at a time")
```
Can't we make it work for both here, to standardize with other processors? (would also avoid duplicated code in this function for images/text)

tonywu71 replied:
In practice, I think the ColPali/ColQwen2 users won't process both text and images in the same time. Anyway, I don't think there is a good way today to combine both text and image processed outputs as explained below.

Currently, both `process_queries` and `process_images` return a `BatchFeature` dict-like object with the following keys:

```python
processor = ColQwen2Processor.from_pretrained("vidore/colqwen2-v1.0-hf")

images = [
    Image.new("RGB", (56, 56), color="white"),
    Image.new("RGB", (32, 32), color="black"),
]
queries = [
    "What is the organizational structure for our R&D department?",
    "Can you provide a breakdown of last year’s financial performance?",
]

processed_images = processor.process_images(images=images)
processed_queries = processor.process_queries(text=queries)

print(processed_queries.keys(), processed_images.keys())
# (dict_keys(['input_ids', 'attention_mask']),
#  dict_keys(['input_ids', 'attention_mask', 'pixel_values', 'image_grid_thw']))
```

I can standardize the processor by making it return a `BatchFeature` dict with all keys. For the `input_ids`, I can concatenate both and pad the tensors. However, since `processed_images["pixel_values"]` is a tensor, I'm really not sure what to put for the i-th row of this tensor.

Even if we solve the previous problem, it's not obvious how `batch_outputs = processor(images=[img_1], text=["str_1", "str_2"])` reorders the processed batched tensor: does `batch_outputs.input_ids[0]` correspond to `img_1` or `str_1`?

Note: I took a look at `CLIPProcessor` and I haven't found my answers as the images processed by CLIP don't have `input_ids`. See below for reference:

```python
import requests
from PIL import Image

from transformers import CLIPProcessor

# model = CLIPModel.from_pretrained("openai/clip-vit-base-patch32")
processor = CLIPProcessor.from_pretrained("openai/clip-vit-base-patch32")


url = "http://images.cocodataset.org/val2017/000000039769.jpg"
image = Image.open(requests.get(url, stream=True).raw)

inputs = processor(
    text=["a photo of a cat", "a photo of a dog"],
    images=image,
    return_tensors="pt",
    padding=True,
)

print(inputs.input_ids.shape)
# torch.Size([2, 7])

print(inputs.pixel_values.shape)
# torch.Size([1, 3, 224, 224])
```

Cyrilvallez replied:
Alright, let's keep it then!

Cyrilvallez commented on tests/models/colqwen2/test_modeling_colqwen2.py:
```diff
@@ -0,0 +1,341 @@
+# coding=utf-8
+# Copyright 2024 The HuggingFace Inc. team. All rights reserved.
```
2025!

Cyrilvallez commented on tests/models/colqwen2/test_processing_colqwen2.py:
```diff
@@ -0,0 +1,246 @@
+import shutil
+import tempfile
```
Missing licence at the top!

Cyrilvallez reviewed (commented):
Hey! Thanks for the clean PR, great work! Left a few comments 🤗
Mostly, I think we can (and should) reinforce the modular part for the modeling itself, as it's mostly similar to ColPali! 👌

tonywu71 commented on docs/source/en/model_doc/colpali.md:
```diff
@@ -54,37 +59,42 @@ images = [
-    "Who printed the edition of Romeo and Juliet?",
     "When was the United States Declaration of Independence proclaimed?",
+    "Who printed the edition of Romeo and Juliet?",
 ]
 
 # Process the inputs
-inputs_images = processor(images=images, return_tensors="pt").to(model.device)
-inputs_text = processor(text=queries, return_tensors="pt").to(model.device)
+inputs_images = processor(images=images).to(model.device)
+inputs_text = processor(text=queries).to(model.device)
```
The change was intended and this makes the quickstart more concise. Having removed `return_tensors="pt"` doesn't change the output of ColPali as the default `return_tensors` is set here:

https://github.com/huggingface/transformers/blob/afc20bc7613d7624e7c7a58486e02268cce86b55/src/transformers/models/colpali/processing_colpali.py#L45

tonywu71 commented on docs/source/en/model_doc/colqwen2.md:
```diff
@@ -0,0 +1,166 @@
+    torch_dtype=torch.bfloat16,
+    device_map="auto",  # "cpu", "cuda", or "mps" for Apple Silicon
+    attn_implementation="flash_attention_2" if is_flash_attn_2_available() else None,
+).eval()
```
Nice, thanks for letting me know! I'll remove the `eval` from the doc then 👍🏼

tonywu71 committed:
docs: remove redundant `.eval()`

tonywu71 commented on src/transformers/models/colpali/modeling_colpali.py:
```diff
@@ -202,47 +205,54 @@ def forward(
-            output[2] = output[2] if output_hidden_states is not None else None
-            output[-1] = (outputs.image_hidden_states if pixel_values is not None else None,)
-            return (loss,) + output if loss is not None else output
+            return tuple(
+                v
+                for v in [
+                    embeddings,
+                    vlm_output.past_key_values,
+                    vlm_hidden_states,
+                    vlm_output.attentions,
+                    vlm_image_hidden_states,
+                ]
+                if v is not None
+            )
```
Wow that's very cool indeed! 😍 I'll change this right away

tonywu71 committed:
refactor:  use `can_return_tuple` decorator for ColPali and ColQwen2

tonywu71 committed:
docs: fix copyright date

tonywu71 committed:
docs: add missing copyright in tests

tonywu71 commented on src/transformers/models/colqwen2/convert_colqwen2_weights_to_hf.py:
```diff
@@ -0,0 +1,208 @@
+    print("Created model with new config and randomly initialized weights")
+
+    # NOTE: The new model was initialized with float32 weights. We need to convert it to the desired precision.
+    # There are two ways to set the model's dtype:
+    # - Using `model.from_pretrained(..., torch_dtype=dtype_precision)` doesn't convert the hyperparameters to the desired precision.
+    # - Using `model.to(dtype_precision)` converts all values - including the hyperparameters - to the desired precision.
```
I did check with @yonigozlan and there was a dtype conversion issue back for `GemmaRotaryEmbedding.inv_freq` when I was working on the ColPali integration in https://github.com/huggingface/transformers/pull/33736:

https://github.com/huggingface/transformers/blob/d19b11f59b0cf337b445827228df796c6c335994/src/transformers/models/colpali/convert_colpali_weights_to_hf.py#L131-L136

But haven't checked this for ColQwen2!

Cyrilvallez replied:
If it's only the RotaryEmbedding, it does not really matter as we force `float32` in the `forward` independently of the dtype of the buffer, to avoid losing precision 😉

tonywu71 replied:
Interesting! I think it's probably a new change because when I was working on https://github.com/huggingface/transformers/pull/33736/files @yonigozlan and I had noticed the predicted embeddings were slightly different without this workaround. If that doesn't change anything, do you mind if we keep it that way?

tonywu71 commented on src/transformers/models/colqwen2/configuration_colqwen2.py:
```diff
@@ -0,0 +1,91 @@
+        self.embedding_dim = embedding_dim
+        super().__init__(**kwargs)
+
+    def get_text_config(self, decoder=False) -> PretrainedConfig:
+        return self.vlm_config
```
It's not a mistake. I was also suprised to see that `Qwen2VLConfig` doesn't have a `text_config` 😅 

https://github.com/huggingface/transformers/blob/4e538409203402058d1ee75c57e03dd908eac508/src/transformers/models/qwen2_vl/configuration_qwen2_vl.py#L59

Cyrilvallez replied:
Not sure what branch this is, but this is not true on `main` anymore, it has a `text_config` 😉

tonywu71 replied:
Yes it's new (see https://github.com/huggingface/transformers/pull/37268)! I'll rebase and make the required changes! 😉

tonywu71 committed:
fix: raise error when `initializer_range` is not in config

tonywu71 commented:
Hi @Cyrilvallez, thanks for the comments! 🤗 

I have addressed most comments, but there are still 2 main questions to answer together:
1. about using modular for the ColQwen2 model → https://github.com/huggingface/transformers/pull/35778#discussion_r2056054263
2. about allowing both text and image inputs for the ColQwen2 processor → https://github.com/huggingface/transformers/pull/35778#discussion_r2055957139.

Happy to discuss about it if needed!

Cyrilvallez commented:
Hey @tonywu71! Even if everything does not fit 1:1 from ColPali, using modular would still allow to cut quite a lot of code in this situation 🤗 So please let's use use it as much as possible! Of course, if things are not compatible, you still need to override them, but in general it will be much simpler 👌

tonywu71 committed:
docs: remove redundant `.eval()` in colpali doc

tonywu71 committed:
fix: fix `get_text_config` now that Qwen2VL has a proper `text_config` attribute

See https://github.com/huggingface/transformers/pull/37268 for details about changes in Qwen2VL's config.

tonywu71 commented:
Hi @Cyrilvallez, thanks for the new review! 🤗  I've addressed a few comments you've raised.

While I totally agree that we cut quite as much redundant code as possible, I'm still convinced that using modular for the ColQwen2 model is not a good idea. Since we are overriding every single method of the parent `ColPaliForRetrieval` class, there is no cut in lines of code. Even worse, we end up with more lines as the code will be in both the modular **and** the modeling files!

Again, this only happens because Qwen2VL is one of the few VLM in `transformers` without a `language_model` attribute and because it has a peculiar way of doing batch sequence packing (hence the `inner_forward` in this PR).

That being said, for future ColVision models that rely on more standard VLMs, we should definitely do as you said and use modular to inherit from ColPali!

Wdyt? Maybe @yonigozlan you could give us your opinion since you've helped for the ColPali implem' too?

tonywu71 committed:
fix: add missing `initializer_range` attribute in `ColQwen2Config`

tonywu71 committed:
fix: use `get_text_config` in `resize_token_embeddings`

yonigozlan committed:
Merge remote-tracking branch 'upstream/main' into add-colqwen2

yonigozlan committed:
update colwen2 with auto_docstring

Cyrilvallez commented on src/transformers/models/colpali/modeling_colpali.py:
```diff
@@ -136,48 +140,42 @@ def forward(
 
-        loss = None
-        if not return_dict:
-            output = (embeddings,) + outputs[2:]
-            output[2] = output[2] if output_hidden_states is not None else None
-            output[-1] = (outputs.image_hidden_states if pixel_values is not None else None,)
-            return (loss,) + output if loss is not None else output
-
         return ColPaliForRetrievalOutput(
-            loss=loss,
             embeddings=embeddings,
```
Why are we removing the loss here? 👀

Cyrilvallez commented on src/transformers/models/colpali/modular_colpali.py:
```diff
@@ -73,10 +73,23 @@ class ColPaliProcessor(PaliGemmaProcessor):
+        image_processor=None,
+        tokenizer=None,
+        chat_template=None,
+        visual_prompt_prefix: str = "Describe the image.",
+        query_prefix: str = "Question: ",
+    ):
+        super().__init__(image_processor=image_processor, tokenizer=tokenizer, chat_template=chat_template)
+        self.visual_prompt_prefix = visual_prompt_prefix
+        self.query_prefix = query_prefix
```
These kind of prefix should be part of the chat_template directly, not hardcoded here 🤗

Cyrilvallez commented on src/transformers/models/colqwen2/convert_colqwen2_weights_to_hf.py:
```diff
@@ -0,0 +1,208 @@
+# coding=utf-8
+# Copyright 2024 The HuggingFace Inc. team.
```
nit: 2025!

Cyrilvallez commented on src/transformers/models/colqwen2/modeling_colqwen2.py:
```diff
@@ -0,0 +1,288 @@
+from .configuration_colqwen2 import ColQwen2Config
+
+
+if is_torch_available():
+    import torch
+    from torch import nn
```
No need to protect the torch import here!

Cyrilvallez commented on src/transformers/models/colqwen2/modeling_colqwen2.py:
```diff
@@ -0,0 +1,288 @@
+    def _init_weights(self, module):
+        std = getattr(self.config, "initializer_range")
+        if std is None:
+            raise AttributeError(
+                "The `initializer_range` attribute is not set in the configuration. Please set it before using the model."
+            )
```
Let's make sure it is correctly defined in the Config with some default value instead of raising here

tonywu71 replied:
The `ColQwen2Config` already has a default value for `initializer_range`, so I'll just remove the `raise` 👍🏼

Cyrilvallez commented on src/transformers/models/colqwen2/modeling_colqwen2.py:
```diff
@@ -0,0 +1,288 @@
+        Uses a custom forward method to fix an issue with the gradient flow when training with multiple GPUs. Code is
+        mostly copied from [`Qwen2VLForConditionalGeneration.forward`], excluding the code for video processing.
+        """
+        if inputs_embeds is None:
+            inputs_embeds = self.vlm.model.embed_tokens(input_ids)
+
+            if pixel_values is not None:
+                pixel_values = pixel_values.type(self.vlm.visual.get_dtype())
+                image_embeds = self.vlm.visual(pixel_values, grid_thw=image_grid_thw)
+                image_mask = (
+                    (input_ids == self.config.vlm_config.image_token_id).unsqueeze(-1).expand_as(inputs_embeds)
+                )
+                image_embeds = image_embeds.to(inputs_embeds.device, inputs_embeds.dtype)
+                inputs_embeds = inputs_embeds.masked_scatter(image_mask, image_embeds)
+
+            if attention_mask is not None:
+                attention_mask = attention_mask.to(inputs_embeds.device)
+
+        outputs = self.vlm.model(
+            input_ids=None,
+            position_ids=position_ids,
+            attention_mask=attention_mask,
+            past_key_values=past_key_values,
+            inputs_embeds=inputs_embeds,
+            use_cache=use_cache,
+            output_attentions=output_attentions,
+            output_hidden_states=output_hidden_states,
+            return_dict=return_dict,
+            cache_position=cache_position,
+        )
+        return outputs
```
If you don't mind, I think it would help readability to have this block directly in the main `forward` instead of separating in 2 functions (due to the large signatures, we need to go back and forth)

tonywu71 replied:
No I don't mind, I think it's actually a good idea! 🤗

Cyrilvallez commented on src/transformers/models/colqwen2/modular_colqwen2.py:
```diff
@@ -0,0 +1,304 @@
+        self.image_token = "<|image_pad|>" if not hasattr(tokenizer, "image_token") else tokenizer.image_token
+        self.video_token = "<|video_pad|>" if not hasattr(tokenizer, "video_token") else tokenizer.video_token
+
+        if visual_prompt_prefix is None:
+            visual_prompt_prefix = "<|im_start|>user\n<|vision_start|><|image_pad|><|vision_end|>Describe the image.<|im_end|><|endoftext|>"
+        self.visual_prompt_prefix = visual_prompt_prefix
+
+        if query_prefix is None:
+            query_prefix = "Query: "
+        self.query_prefix = query_prefix
```
These should be incorporated to the chat_template 🤗

yonigozlan replied:
Not sure if we should have a chat template here since this is not a chat model really. We had the same issue with Got OCR and ended up not using a chat template. wdyt?

Cyrilvallez commented on src/transformers/models/colqwen2/modular_colqwen2.py:
```diff
@@ -0,0 +1,304 @@
+        **kwargs: Unpack[ColQwen2ProcessorKwargs],
+    ) -> BatchFeature:
+        """
+        Prepare for the model one or several image(s). This method is a wrapper around the `__call__` method of the ColQwen2Processor's
+        [`ColQwen2Processor.__call__`].
+
+        This method forwards the `images` and `kwargs` arguments to Qwen2VLImageProcessor's [`~Qwen2VLImageProcessor.__call__`].
+
+        Args:
+            images (`PIL.Image.Image`, `np.ndarray`, `torch.Tensor`, `List[PIL.Image.Image]`, `List[np.ndarray]`, `List[torch.Tensor]`):
+                The image or batch of images to be prepared. Each image can be a PIL image, NumPy array or PyTorch
+                tensor. In case of a NumPy array/PyTorch tensor, each image should be of shape (C, H, W), where C is a
+                number of channels, H and W are image height and width.
+            return_tensors (`str` or [`~utils.TensorType`], *optional*):
+                If set, will return tensors of a particular framework. Acceptable values are:
+
+                - `'tf'`: Return TensorFlow `tf.constant` objects.
+                - `'pt'`: Return PyTorch `torch.Tensor` objects.
+                - `'np'`: Return NumPy `np.ndarray` objects.
+                - `'jax'`: Return JAX `jnp.ndarray` objects.
+
+        Returns:
+            [`BatchFeature`]: A [`BatchFeature`] with the following fields:
+
+            - **input_ids** -- List of token ids to be fed to a model.
+            - **attention_mask** -- List of indices specifying which tokens should be attended to by the model (when
+              `return_attention_mask=True` or if *"attention_mask"* is in `self.model_input_names` and if `text` is not
+              `None`).
+            - **pixel_values** -- Pixel values to be fed to a model. Returned when `images` is not `None`.
+        """
+        return self.__call__(images=images, **kwargs)
+
+    def process_queries(
+        self,
+        text: Union[TextInput, List[TextInput]],
+        **kwargs: Unpack[ColQwen2ProcessorKwargs],
+    ) -> BatchFeature:
+        """
+        Prepare for the model one or several texts. This method is a wrapper around the `__call__` method of the ColQwen2Processor's
+        [`ColQwen2Processor.__call__`].
```
Why do we redefine them here? They will be inherited directly!

tonywu71 replied:
Oh you're right! However, I think the docstring will inherit from `ColPaliProcessor`'s docstring and thus referencing ColPali. Is there a way to simply override the docstring here? If not, should we keep the code as it is?

tonywu71 replied:
Wdyt @yonigozlan?

yonigozlan replied:
I don't see a clean way to do this, but we can just remove specific references to the tokenizer and image processor in the docstring imo

Cyrilvallez reviewed (commented):
Hey! Sorry for the delay! This is pretty clean, great work! 🤗 I just left a few last comments!

tonywu71 committed:
docs: fix wrong copyright year

tonywu71 committed:
chore: remove `raise` as `initializer_range` has a default value in `ColQwen2Config`

tonywu71 committed:
refactor: merge `inner_forward` into `forward`

yonigozlan committed:
Merge remote-tracking branch 'upstream/main' into add-colqwen2

yonigozlan committed:
Refactor colqwen2 after refactoring of qwen2VL, use modular for modeling code

yonigozlan commented:
@Cyrilvallez taking over for the final push on this PR as Tony is quite busy. I pushed some necessary updates after the refactoring of Qwen2VL (so nice to have btw), all should be good now and we use modular much more, including for the modeling code 🤗. @tonywu71 you'll still have to run the updated convert_weights script and push to your repo :), but apart from that we should be ready to merge!

yonigozlan committed:
protect torch import in modular to protect in processing

yonigozlan committed:
protect torch import in modular to protect in processing

yonigozlan committed:
Merge branch 'add-colqwen2' of https://github.com/tonywu71/transformers into add-colqwen2

tonywu71 committed:
tests: fix hf model path in ColQwen2 integration test

tonywu71 commented:
> @Cyrilvallez taking over for the final push on this PR as Tony is quite busy. I pushed some necessary updates after the refactoring of Qwen2VL (so nice to have btw), all should be good now and we use modular much more, including for the modeling code 🤗. @tonywu71 you'll still have to run the updated convert_weights script and push to your repo :), but apart from that we should be ready to merge!

@yonigozlan Done, the model repo is updated! 🤗 I've also pushed a commit to fix the Hf model path for ColQwen2 integration tests. Lmk if there's anything left to do before merging!

Cyrilvallez commented on docs/source/en/model_doc/colqwen2.md:
```diff
@@ -0,0 +1,166 @@
+    model_name,
+    torch_dtype=torch.bfloat16,
+    device_map="auto",  # "cpu", "cuda", or "mps" for Apple Silicon
+    attn_implementation="flash_attention_2" if is_flash_attn_2_available() else None,
+)
```
This is a super nit, feel free to disregard if you're too annoyed by the review process 😆 But passing `None` is a bit misleading for an example IMO, even if it's equivalent
```suggestion
    attn_implementation="flash_attention_2" if is_flash_attn_2_available() else "sdpa",
)
```

tonywu71 replied:
Actually it seems `sdpa` doesn't work out-of-the-box for ColQwen2 as I get this error when loading the model on MPS.

❌ Code:

```python
model_name = "vidore/colqwen2-v1.0-hf"

# Load model
model = ColQwen2ForRetrieval.from_pretrained(
    model_name,
    torch_dtype=torch.bfloat16,
    device_map="auto",  # "cpu", "cuda", or "mps" for Apple Silicon
    attn_implementation="flash_attention_2" if is_flash_attn_2_available() else "sdpa",
)
```

The error:

```
ValueError: ColQwen2ForRetrieval does not support an attention implementation through torch.nn.functional.scaled_dot_product_attention yet. Please request the support for this architecture: https://github.com/huggingface/transformers/issues/28005. If you believe this error is a bug, please open an issue in Transformers GitHub repository and load your model with the argument `attn_implementation="eager"` meanwhile. Example: `model = AutoModel.from_pretrained("openai/whisper-tiny", attn_implementation="eager")`
```

✅ However, I managed to load Qwen2VL with SDPA:

```python
model = Qwen2VLForConditionalGeneration.from_pretrained(
    "Qwen/Qwen2-VL-2B-Instruct",
    torch_dtype=torch.bfloat16,
    device_map="auto",  # "cpu", "cuda", or "mps" for Apple Silicon
    attn_implementation="flash_attention_2" if is_flash_attn_2_available() else "sdpa",
)
```

@Cyrilvallez @yonigozlan I read about [the instructions for enabling SDPA](https://github.com/huggingface/transformers/issues/28005) on ColQwen2 but next steps are a bit unclear as ColQwen2 essentially piggybacks on Qwen2VL thanks to modular. Any ideas about the right fix? 🤗

Cyrilvallez replied:
I believe it's only because the flags are not set in the PreTrainedModel - adding 

```python
_supports_flash_attn_2 = True
_supports_sdpa = True
_supports_flex_attn = True
_supports_cache_class = True
```

should solve it

tonywu71 replied:
Tsm, the fix is working like a charm!

Cyrilvallez reviewed (approved):
All right! Amazing work, congrats to you both @tonywu71 @yonigozlan! Super clean 🤗 I left 2 ultra small comments as my job here is to be very picky 🙃, but that's it! Feel free to merge @yonigozlan! 
Thanks for the great addition 🤗

tonywu71 committed:
docs: clarify `attn_implementation` and add comments

tonywu71 committed:
docs: add fallback snippet for using offline PIL dummy images

tonywu71 committed:
docs: temporarily revert attn_implementation to `None` while sdpa is not fixed

tonywu71 committed:
docs: tweaks in colpali/colqwen2 quick start snippets

tonywu71 committed:
fix: add missing flags to enable SDPA/Flex Attention in ColQwen2 model

tonywu71 committed:
fix: add missing changes in modular file

yonigozlan committed:
Merge remote-tracking branch 'upstream/main' into add-colqwen2

yonigozlan committed:
fix modeling tests
